src/mgr/DaemonServer.cc

```diff
@@ -2983,13 +2983,19 @@ void DaemonServer::got_service_map()
       if (pending_service_map.epoch == 0) {
 	// we just started up
 	dout(10) << "got initial map e" << service_map.epoch << dendl;
+	ceph_assert(pending_service_map_dirty == 0);
+	pending_service_map = service_map;
+	pending_service_map.epoch = service_map.epoch + 1;
+      } else if (pending_service_map.epoch <= service_map.epoch) {
```

Comment:
I am not sure I got your question right. Anyway trying to clarify.

When the mgr is activated (`pending_service_map.epoch == 0`), we receive the service map from a mon, use it to initialize `pending_service_map` and set `pending_service_map.epoch` to the current epoch + 1. Normally, after this a new service map update will go from this mgr to the mon, and when sending it the mgr always increases its `pending_service_map.epoch`, so when the map gets back committed from the mon we may always expect `pending_service_map.epoch > service_map.epoch`.

But there is a case on failover, when the previous active mgr sends two service_map service map updates before shutdown, and the mon sends them both to the new active mgr. You may find more details about such a case in the tracker ticket. So in this situation, when receiving the first service map (`epoch == x`) we will set `pending_service_map.epoch = x + 1` , end then when receive the second map (`epoch == x + 1`) we will have a condition `pending_service_map.epoch == service_map.epoch`. And this is what I want to handle here.

Actually, the condition block `if (pending_service_map.epoch == 0)` is redundant here now, because it could also be handled by the `if (pending_service_map.epoch <= service_map.epoch)` condition block. I left it only, as it seems more clear what it going on there, because we are always expected to hit `if (pending_service_map.epoch == 0)` condition and only very occasionally the second one.

Reply:
Thanks for the additional clarification. The fix makes sense to me.